provider/akamai.go

```diff
@@ -0,0 +1,413 @@
+	return c.Do(config, req)
+}
+
+// AkamaiConfig clarifies the method signature
+type AkamaiConfig struct {
+	DomainFilter          DomainFilter
+	ZoneIDFilter          ZoneIDFilter
+	ServiceConsumerDomain string
+	ClientToken           string
+	ClientSecret          string
+	AccessToken           string
+	DryRun                bool
+}
+
+// AkamaiProvider implements the DNS provider for Akamai.
+type AkamaiProvider struct {
+	domainFilter DomainFilter
+	zoneIDFilter ZoneIDFilter
+	config       edgegrid.Config
+	dryRun       bool
+	client       akamaiClient
+}
+
+type akamaiError struct {
+	Title     string `json:"title"`
+	Status    int    `json:"status"`
+	Detail    string `json:"detail"`
+	RequestID string `json:"requestId"`
+}
+
+type akamaiZones struct {
+	Zones []akamaiZone `json:"zones"`
+}
+
+type akamaiZone struct {
+	ContractID string `json:"contractId"`
+	Zone       string `json:"zone"`
+}
+
+type akamaiRecordsets struct {
+	Recordsets []akamaiRecord `json:"recordsets"`
+}
+
+type akamaiRecord struct {
+	Name  string        `json:"name"`
+	Type  string        `json:"type"`
+	TTL   int64         `json:"ttl"`
+	Rdata []interface{} `json:"rdata"`
+}
+
+// NewAkamaiProvider initializes a new Akamai DNS based Provider.
+func NewAkamaiProvider(akamaiConfig AkamaiConfig) *AkamaiProvider {
+	edgeGridConfig := edgegrid.Config{
+		Host:         akamaiConfig.ServiceConsumerDomain,
+		ClientToken:  akamaiConfig.ClientToken,
+		ClientSecret: akamaiConfig.ClientSecret,
+		AccessToken:  akamaiConfig.AccessToken,
+		MaxBody:      1024,
+		HeaderToSign: []string{
+			"X-External-DNS",
+		},
+		Debug: false,
+	}
+
+	provider := &AkamaiProvider{
+		domainFilter: akamaiConfig.DomainFilter,
+		zoneIDFilter: akamaiConfig.ZoneIDFilter,
+		config:       edgeGridConfig,
+		dryRun:       akamaiConfig.DryRun,
+		client:       &akamaiOpenClient{},
+	}
+	return provider
+}
+
+func (p *AkamaiProvider) request(method, path string, body io.Reader) ([]byte, error) {
+	req, err := p.client.NewRequest(p.config, method, fmt.Sprintf("https://%s/%s", p.config.Host, path), body)
+	if err != nil {
+		log.Errorf("Akamai client failed to prepare the request")
+		return nil, err
+	}
+	resp, err := p.client.Do(p.config, req)
+	if err != nil {
+		log.Errorf("Akamai client failed to do the request")
+		return nil, err
+	}
+
+	//204 means no body on success by choice
+	if resp.StatusCode == 204 {
+		return nil, nil
+	}
+
+	defer resp.Body.Close()
+	byt, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Errorf("ioutil failed to read the response-body of the request to Akamai")
+		return nil, err
+	}
+
+	//catch authentication errors here, they aren't handled very well by the client
+	akamaiError := akamaiError{}
+	err = json.Unmarshal([]byte(byt), &akamaiError)
+	if err != nil {
+		log.Errorf("Failed to unmarshal the response-body of the request to Akamai")
+		return nil, err
+	}
+	if akamaiError.Status >= 300 {
```

Comment:
Sadly the Akamai go client isn't returning errors in the way I'd expect it to.
The 'Do' function does not inform in any way if there is a failure with for example authentication.
It just gives back a json stream and nil.
If I use wrong authentication data, the Akamai client answers with a 200 statuscode and a jsonbody with such data for example:
```
title: Bad request 
status: 400 
detail: Invalid authorization client token 
requestId: b8280332
```
So I am quite unsure how to handle all errors that could happen correctly.
I'd opt for at least '>= 400' because of such erroneous events that are being invoked by the user.
I chose '300' for events where Akamai changes it's API and starts answering to our requests with a '301  Moved Permanently' for example. I don't think there is another way of informing the user of such error.
But i am totally open for suggestions, what do you think?

Reply:
I had another look at the akamai client and found that they internally do almost the same.
I now use their function 'IsSuccess' which basically is only resolves to true if the statuscode is within the range of the 200 http codes. This should get all errors.